hibike/hibike_simulator.py

```diff
@@ -1,133 +1,156 @@
-            self.subscribedTo[UID]['is_low'] = not self.subscribedTo[UID]['is_low']
-        elif should_update:
-            self.subscribedTo[UID]['time'] = curr_time
-            self.subscribedTo[UID]['data'] = self.__getRandomData(UID, not self.subscribedTo[UID]['is_low'])
-        return self.subscribedTo[UID]['data'] #returns sensor data
-
-    def __calculateFlipTime(self, UID):
-        """ Determines time of flip. """
-
-        rand, last_time = random.random(), self.subscribedTo[UID]['time']
-        noise = random.random()
-        for i in range(len(cutoffs)):
-            if rand < cutoffs[i]:
-                return (time.time() + noise + i)*1000
-
-    def __getRandomData(self, UID, is_low):
-        """ Finds device_type of UID and returns corresponding flipped data."""
-
-        device_type = UID[:6]
-        low, high, noise = sensor_values[device_type]
-        if low is None or high is None or noise is None:
-            return 0 # TODO: what to do with these device types?
-        if is_low:
-            return high + (random.random() - .5) * noise
+import threading
+import multiprocessing
+import queue
+
+from hibike_util import *
+
+fake_uids = [0 << 72, 7 << 72]
+
+# This is what actually "knows" about. Only updated after enumeration
+uid_to_queue = {}
+
+
+def hibike_process(badThingsQueue, stateQueue, pipeFromChild):
+    devices = [FakeDevice(uid, stateQueue) for uid in fake_uids]
+    for device in devices:
+        device.start()
+
+    # the main thread reads instructions from statemanager and forwards them to the appropriate device write threads
+    while True:
+        instruction, args = pipeFromChild.recv()
+        if instruction == "enumerate_all":
+            for device in devices:
+                device.instruction_queue.put(("ping", []))
+        elif instruction == "subscribe_device":
+            uid = args[0]
+            if uid in uid_to_queue:
+                uid_to_queue[uid].put(("subscribe", args))
+            else:
+                print('not found uid')
+
+def runDeviceWrite(instruction_queue, fake_device_queue):
+    # instruction_queue is input
+    # fake_device_queue is output to smart sensor
+    while True:
+        instruction, args = instruction_queue.get()
+
+        if instruction == "ping":
+            fake_device_queue.put(0)
+        elif instruction == "subscribe":
+            uid, delay, params = tuple(args)
+            fake_device_queue.put(delay)
+            fake_device_queue.put(params)
+
+def runDeviceRead(fake_device_readings, stateQueue, instruction_queue):
+    while True:
+        result, values = fake_device_readings.get()
+        res = None
+        if result == "device_values":
+            res = [result, values]
+        elif result == "device_subscribed":
+            uid = values[0]
+            if uid in uid_to_queue:
+                del uid_to_queue[uid]
+            uid_to_queue[uid] = instruction_queue
+            res = ["device_subscribed", values]
```

Comment:
Also, it seems odd that you're making `res` a list and not a tuple in this function